cmd/cstor-pool-mgmt/controller/replica-controller/new_replica_controller.go

```diff
@@ -59,119 +60,178 @@ type CStorVolumeReplicaController struct {
-				controller.recorder.Event(newCVR, corev1.EventTypeNormal, string(common.SuccessSynced), string(common.MessageDestroySynced))
-			} else {
-				q.Operation = common.QOpModify
-				glog.Infof("cStorVolumeReplica Modify event : %v, %v", newCVR.ObjectMeta.Name, string(newCVR.ObjectMeta.UID))
-				controller.recorder.Event(newCVR, corev1.EventTypeNormal, string(common.SuccessSynced), string(common.MessageModifySynced))
-				return // will be removed once modify is implemented
-			}
-			controller.enqueueCStorReplica(newCVR, q)
-		},
-		DeleteFunc: func(obj interface{}) {
-			cVR := obj.(*apis.CStorVolumeReplica)
-			if !IsRightCStorVolumeReplica(cVR) {
-				return
-			}
-			glog.Infof("cVR Resource deleted event: %v, %v", cVR.ObjectMeta.Name, string(cVR.ObjectMeta.UID))
-		},
-	})
+	glog.Info("will set up informer event handlers for cvr")
+
+	ql := common.QueueLoad{}
+
+	cvrInformer.Informer().
+		AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				cvrObj := obj.(*apis.CStorVolumeReplica)
+
+				glog.V(4).Infof(
+					"received informer add event for cvr {%s}",
+					cvrObj.Name,
+				)
+
+				if !IsRightCStorVolumeReplica(cvrObj) ||
+					IsDeletionFailedBefore(cvrObj) ||
+					IsErrorDuplicate(cvrObj) {
+					return
+				}
+
+				ql.Operation = common.QOpAdd
+				controller.recorder.Event(
+					cvrObj,
+					corev1.EventTypeNormal,
+					string(common.SuccessSynced),
+					string(common.MessageCreateSynced),
+				)
+
+				// new cvr requests will have phase as blank
+				//
+				// NOTE:
+				//  for every restart of controller container
+				// this informer handler will get add event
+				// for each cvr resource present in k8s
+				if IsEmptyStatus(cvrObj) {
+					cvrObj.Status.Phase = apis.CVRStatusInit
+				} else {
+					cvrObj.Status.Phase = apis.CVRStatusRecreate
+				}
+
+				cvrObj, _ = controller.clientset.
+					OpenebsV1alpha1().
+					CStorVolumeReplicas(cvrObj.Namespace).
+					Update(cvrObj)
+
+					// push this operation to workqueue
+				controller.enqueueCStorReplica(cvrObj, ql)
+			},
+
+			UpdateFunc: func(old, new interface{}) {
+				newCVR := new.(*apis.CStorVolumeReplica)
+				oldCVR := old.(*apis.CStorVolumeReplica)
+
+				glog.V(4).Infof(
+					"received informer update event for cvr {%s}",
+					newCVR.Name,
+				)
+
+				if !IsRightCStorVolumeReplica(newCVR) ||
+					IsErrorDuplicate(newCVR) ||
+					IsOnlyStatusChange(oldCVR, newCVR) {
+					return
+				}
+
+				// TODO: Need to have status in more organised manner
+
+				// periodic resync will send update events for all known cvr
+				if newCVR.ResourceVersion == oldCVR.ResourceVersion {
+					ql.Operation = common.QOpSync
+
+					controller.recorder.Event(
+						newCVR,
+						corev1.EventTypeNormal,
+						string(common.SuccessSynced),
+						string(common.StatusSynced),
+					)
+
+				} else if IsDestroyEvent(newCVR) {
```

Comment:
is it good if we reverse this and the above 'if', i.e., how about below order:
```
if IsDestroyEvent(newCVR) {
} else if newCVR.ResourceVersion == oldCVR.ResourceVersion{
}else {
}
```

Reply:
what is the motivation for above? @vishnuitta

Reply:
Understood. Done